### ISSUE TITLE
Better support for 64-bit #defines

### ISSUE THREAD

**Issue by Timmmm** [swig/swig#1287]
At the moment if you want to do something like this

    #define FOO 0xdeadbeefdeadbeef

Unfortunately it will generate code that truncates it to an `int`. (I'm using Python, but I assume it is similar for other languages). Python supports 64-bit integers so this should be possible. You can do:

    #define FOO (unsigned long long) 0xdeadbeefdeadbeef

But then it ignores the constant entirely.

A solution is:

    %constant unsigned long long FOO = 0xdeadbeefdeadbeef;

However that means you can't easily give it the same name as the `#define`'d constant. Also note that `uint64_t` does something different to `unsigned long long`.

This is all rather awkward. I think the nicest solution would be to provide a way to override the detected type of `#define`'s in the `.i` file, so you have

     #define FOO 0xdeadbeefdeadbeef

In your `foo.h` file, and then

    %include "foo.h"
    %constant unsigned long long FOO;

In your `foo.i` file.

**Comment by wsfulton** [comment]
I suggest you use the C ULL suffix to work around this:
```c
#define FOO 0xdeadbeefdeadbeefULL
```
SWIG will then recognise this as an unsigned long long C type and marshall it using a python 2 long or python 3 int type.

Ultimately SWIG should recognise that this is a C unsigned long long type and this is a bug that needs fixing.

**Comment by ojwb** [comment]
Still reproducible with git master.

> Ultimately SWIG should recognise that this is a C unsigned long long type and this is a bug that needs fixing.

An awkward obstacle to that is that at SWIG-time we don't know the type widths.  E.g. if the generated wrapper is compiled on x86-64 Linux then `0xdeadbeefdeadbeef` fits in an `unsigned long`, but on some other platforms it needs to be `unsigned long long`.  I guess we can either be conservative (pick the smallest type that C/C++ guarantees will fit), or we use the types with widths like `uint64_t` (which aren't available in older standards versions), or generate conditional code.

**Comment by Timmmm** [comment]
> which aren't available in older standards versions

How old are you talking? I'm guessing if anyone is still using C89 or C++03 then they probably aren't going to be using a version of SWIG from 2022!

**Comment by ojwb** [comment]
ISO C90 and C++98 are the current baselines.  The aim is to support newer standards when the compiler supports them but support for a feature in C++nn shouldn't require C++mm where mm > nn.

I don't know at what point we drop support for older C and C++, but we haven't done so yet.  William tends to be ultra-conservative about such things.

**Comment by leonardgerardatomicmachinescom** [comment]
This bit me pretty hard and is probably generating bugs silently in many code base. C++ literal with no suffix is supposed to be understood as having the smallest type able to fit it https://en.cppreference.com/w/cpp/language/integer_literal 

Where in the code could this be fixed? (If one wants to try to write a patch).

**Comment by ojwb** [comment]
It'd need to be in or before we get to the parser (`Source/CParse/parser.y`) since that needs to have type codes to be able to deduce types for C++ `auto`, `decltype`, etc.

The tricky part is picking the type.  A C/C++ compiler easily can because it knows the sizes of the types for the target architecture, but as I pointed out above SWIG aims to generate code that can then be compiled (and work!) on any platform.

Note we can't just use e.g. `uint64_t` for `0xdeadbeefdeadbeef` for several reasons.  First problem is that is the wrong type - as the page you link to says, it should be one of  `int`, `unsigned int`, `long int`, `unsigned long int`, `long long int` (since C++11), `unsigned long long int` (since C++11).  Another issue is that SWIG still aims to generate code that works on C90 and C++98 unless either we're wrapping a feature from a newer standard or the target language C or C++ extension API has a particular minimum standards version requirement (and if you don't agree this makes sense in 2024, I'm really not the person you need to convince).

I think it would be reasonable to ignore platforms which are theoretical (e.g. the standard allows `int` to be 128 bits, but I don't think any implementation actually is) or obsolete (e.g. whichever old platform it was that had 32-bit `char`, `int` and `long`, so `sizeof(long)` was 1) and only worry about platforms that actually exist and are in current use (and which a C and/or C++ compiler targets and at least one SWIG target language runs on).

So one approach might be to pick a type which is wide enough to hold the constant on any platform (ignoring theoretical, etc as above).  However a third and perhaps less obvious problem is that some of the target languages have a limited set of integer types compared to C/C++, and the sizes of those types can vary by platform so by conservatively picking a type that we then can't represent in the target language we could break cases which actually work fine currently.

It's certainly bad that such constants are quietly mishandled.  I suspect fixing just the quiet mishandling would be much simpler - we know the constant and the type we've chosen so we could generate code that failed to compile if the constant didn't fit in that chosen type, perhaps:

```
/* Comment giving some advice on how to work around this... */
typedef int swig_constant_type_assertion[(int)0xdeadbeefdeadbeef) != 0xdeadbeefdeadbeef ? -1 : 1];
```

**Comment by leonardgerardatomicmachinescom** [comment]
I agree that we don't want to break things which are working today. I am unsure how we can craft such a case though.

Let's assume swig made all the #defines of integers as long (or long long if we allow C99, but then we could just use stdint.h ...). Do you have an example of something that would break?


In any case, I really like your proposal of relying on the compiler to do the final checking, and fail loudly when swig picked the wrong type!

**Comment by ojwb** [comment]
> Let's assume swig made all the #defines of integers as long. Do you have an example of something that would break?

Sorry, seems I missed this question before.

Deliberately using the wrong type for constants opens a can of worms when C++ type deduction is involved because that allows that wrong type to be copied to essentially arbitrary other places, such as function parameters - e.g.

```
#define C_INT 1234

void f(decltype(C_INT)&);
```

This should effectively mean `void f(int&);` but if SWIG makes `C_INT` type `long` instead if `int` then it will interpret this as `void f(long&);` instead.

I haven't constructed a full example showing breaking from that, but hopefully it's clear this is not a sensible path to be heading down - we want to be fixing existing cases of SWIG using incorrect types, not widening the cases in which it uses incorrect types.

So better would be to teach SWIG about the rule that it should use a wider type for an integer constant when it needs to.  The tricky part (as mentioned above) is that SWIG aims to generate portable code so we don't know the type widths, aside from what the C/C++ standards require of them (so `long` is at least as wide as `int`, `short` is at least 16 bits, etc).

A fairly simple change which isn't perfect but I think only improves things (at least for real-world platforms) would be to treat an unsuffixed constant which doesn't fit in a `int` as `long`.  The original example above of `0xdeadbeefdeadbeef` would then be `long` and that would at least work for platforms where `long` is 64-bits (so 64-bit Linux, and most other 64-bit Unix-like platforms).  This wouldn't help where `sizeof(long) == sizeof(int)` but neither does assuming `int` like we currently do.
